src/heaps/binary_heap.jl

```diff
@@ -1,124 +1,41 @@
-                if compare(comp, rv, v)
-                    @inbounds valtree[i] = rv
-                    i = rc
-                else
-                    swapped = false
-                end
-            else
-                if compare(comp, lv, v)
-                    @inbounds valtree[i] = lv
-                    i = lc
-                else
-                    swapped = false
-                end
-            end
-        else        # contains only left child
-            @inbounds lv = valtree[lc]
-            if compare(comp, lv, v)
-                @inbounds valtree[i] = lv
-                i = lc
-            else
-                swapped = false
-            end
-        end
-    end
-
-    valtree[i] = v
-end
-
-
-function _binary_heap_pop!(comp::Comp, valtree::Array{T}) where {Comp,T}
-    # extract root
-    v = valtree[1]
-
-    if length(valtree) == 1
-        empty!(valtree)
-    else
-        valtree[1] = pop!(valtree)
-        if length(valtree) > 1
-            _heap_bubble_down!(comp, valtree, 1)
-        end
-    end
-    v
-end
-
-
-function _make_binary_heap(comp::Comp, ty::Type{T}, xs) where {Comp,T}
-    n = length(xs)
-    valtree = copy(xs)
-    for i = 2 : n
-        _heap_bubble_up!(comp, valtree, i)
-    end
-    valtree
-end
-
+include("arrays_as_heaps.jl")
 
 #################################################
 #
 #   heap type and constructors
 #
 #################################################
 
-mutable struct BinaryHeap{T,Comp} <: AbstractHeap{T}
-    comparer::Comp
+#=
+These structs may be substituted by Base.Forward and Base.Reverse,
+but float comparison will be 2x slower to preserve ordering with NAN values.
+=#
+struct FasterForward <: Base.Ordering end
+struct FasterReverse <: Base.Ordering end
+Base.lt(o::FasterForward, a, b) = a < b
+Base.lt(o::FasterReverse, a, b) = a > b
+
+mutable struct BinaryHeap{T, O <: Base.Ordering} <: AbstractHeap{T}
     valtree::Vector{T}
+    ordering::O
 
-    BinaryHeap{T,Comp}() where {T,Comp} = new{T,Comp}(Comp(), Vector{T}())
+    # min heap by default
+    function BinaryHeap(::Type{T}, ordering::O = FasterForward()) where {T,O}
+        new{T,O}(Vector{T}(), ordering)
+    end
 
-    function BinaryHeap{T,Comp}(xs::AbstractVector{T}) where {T,Comp}
-        valtree = _make_binary_heap(Comp(), T, xs)
-        new{T,Comp}(Comp(), valtree)
+    function BinaryHeap(xs::AbstractVector{T}, ordering::O = FasterForward()) where {T,O}
+        valtree = heapify(xs, ordering)
+        new{T,O}(valtree, ordering)
     end
 end
 
-const BinaryMinHeap{T} = BinaryHeap{T, LessThan}
-const BinaryMaxHeap{T} = BinaryHeap{T, GreaterThan}
-
-BinaryMinHeap(xs::AbstractVector{T}) where T = BinaryMinHeap{T}(xs)
-BinaryMaxHeap(xs::AbstractVector{T}) where T = BinaryMaxHeap{T}(xs)
+BinaryMinHeap(xs::AbstractVector) = BinaryHeap(xs, FasterForward())
+BinaryMaxHeap(xs::AbstractVector) = BinaryHeap(xs, FasterReverse())
+BinaryMinHeap(::Type{T}) where T = BinaryHeap(T, FasterForward())
```

Comment:
Do we need to take the type as an argument?
I believe we can take it as a type-parameter as in:
```
julia> struct Foo{T}
       x::Vector{T}
       end

julia> Foo{T}(n::Int) where T = Foo{T}(ones(n))

julia> Foo{UInt8}(10)
Foo{UInt8}(UInt8[0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01])
```

so we can just do:
```suggestion
BinaryMinHeap{T}() where T = BinaryHeap(T, FasterForward())
```

Except we should probably replace that one too so it would be:
```suggestion
BinaryMinHeap{T}() where T = BinaryHeap{T}(FasterForward())
```

Reply:
If we allow a type parameter and input vector, then mismatched types will trigger an implicit conversion and performance penalty. I think it's better to force explicit conversion by only allowing either type or vector. I also think it's nicer for each of these pieces of information (type or vector) to occupy the same argument position. Open to other API suggestions though.
```
# Baseline performance with matching types

julia> @btime Foo{Float64}(ones(3)) # ones defaults to Float64
  29.514 ns (2 allocations: 128 bytes)
Foo{Float64}([1.0, 1.0, 1.0])

# Performance penalty to convert mismatched types

julia> @btime Foo{Int}(ones(3))
  59.776 ns (3 allocations: 240 bytes)
Foo{Int64}([1, 1, 1])

# Explicit conversion performance is no worse than implicit conversion, but intent is move obvious.

julia> @btime Foo{Int}(convert(Vector{Int}, ones(3)))
  59.305 ns (3 allocations: 240 bytes)
Foo{Int64}([1, 1, 1])
```

Reply:
we should be consistent with the API of `Base`.
Base allows this, we should also.
Base made the call that allowing this implict conversion was acceptable.
We should to.


```
julia> Vector{Float64}([1,0])
2-element Array{Float64,1}:
 1.0
 0.0
```


Its <2x overhead and if you are 
constructing new heaps in your critical loop then that is its own problem.

Further we also (again inline with  Base), we can push! the wrong typpe and get conversion. Which I feel is a useful feature.

Reply:
Are we now inline with `Base` ?

Reply:
> Are we now inline with `Base` ?

Implicit conversion is enabled in latest pushed commit: https://github.com/JuliaCollections/DataStructures.jl/pull/547/commits/1eeafc81efea37453b7e068f033b276e8ba66ae9